kernel/src/mem.rs

```diff
@@ -413,20 +463,359 @@ impl Read for ReadOnlyAppSlice {
         }
     }
 
-    fn map_or<F, R>(&self, default: R, fun: F) -> R
+    fn enter<F, R>(&self, fun: F) -> Result<R, process::Error>
     where
-        F: FnOnce(&[u8]) -> R,
+        F: FnOnce(&ReadableProcessSlice) -> R,
     {
         match self.process_id {
-            None => default,
-            Some(pid) => pid.kernel.process_map_or(default, pid, |_| {
-                // Safety: `kernel.process_map_or()` validates that the process still exists
-                // and its memory is still valid. `Process` tracks the "high water mark" of
-                // memory that the process has `allow`ed to the kernel, and will not permit
-                // the process to free any memory after it has been `allow`ed. This guarantees
-                // that the buffer is safe to convert into a slice here.
-                fun(unsafe { raw_appslice_to_slice(self.ptr, self.len) })
-            }),
+            None => Err(process::Error::NoSuchApp),
+            Some(pid) => pid
+                .kernel
+                .process_map_or(Err(process::Error::NoSuchApp), pid, |_| {
+                    // Safety: `kernel.process_map_or()` validates that
+                    // the process still exists and its memory is still
+                    // valid. `Process` tracks the "high water mark" of
+                    // memory that the process has `allow`ed to the
+                    // kernel, and will not permit the process to free any
+                    // memory after it has been `allow`ed. This guarantees
+                    // that the buffer is safe to convert into a slice
+                    // here.
```

Comment:
```suggestion
                    // Safety: `kernel.process_map_or()` validates that
                    // the process still exists and its memory is still
                    // valid. Once a `ProcessBuffer` is created in a process
                    // there is no API for removing it, so any `Process`
                    // implementation must assume the ProcessBuffer always
                    // exists and will not permit the process to free any
                    // memory after it has been `allow`ed. This guarantees
                    // that the buffer is safe to convert into a slice
                    // here.
```

Reply:
I understand the motivation behind this change, and this seems correct to me. However, I fear that the rephrased comment might be more confusing given it does not clearly express the relation between these safety considerations and the process' high water mark any longer. Specifically, I think that "must assume the ProcessBuffer always exists" could be confusing as, from a capsule developer perspective, this struct is passed back to the kernel on a call to `allow_read{only,write}`. Do you mind if I try rephrasing?

Reply:
Yeah I have no attachment to this text. What I would like to avoid is tying the safety of process buffers to a particular implementation detail of `trait Process` as there is no guarantee that process buffers will only be used with ProcessStandard.